File: src/main/java/red/tetracube/data/entities/AuthenticationToken.java
package red.tetracube.data.entities;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.Calendar;
import java.util.UUID;

@Entity
@Table(schema = "gateway", name = "authentication_tokens")
public class AuthenticationToken {

    @Id
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "token", nullable = false, unique = true)
    private String token;

    @Column(name = "valid_from", nullable = false)
    private Timestamp validFrom;

    @Column(name = "valid_until", nullable = false)
    private Timestamp validUntil;

    @Column(name = "is_valid", nullable = false)
    private Boolean isValid;

    @Column(name = "in_use", nullable = false)
    private Boolean inUse;

    @JoinColumn(name = "id_house", nullable = false)
    @ManyToOne(fetch = FetchType.LAZY, targetEntity = House.class)
    private House house;

    public AuthenticationToken() {
    }

    public AuthenticationToken(String token, House house) {
        var now = Calendar.getInstance();
        var validUntil = Calendar.getInstance();
        validUntil.add(Calendar.YEAR, 1);
        this.id = UUID.randomUUID();
        this.token = token;
        this.validFrom = new Timestamp(now.getTimeInMillis());
        this.validUntil = new Timestamp(validUntil.getTimeInMillis());
        this.isValid = true;
        this.inUse = false;
        this.house = house;
    }

    public String getToken() {
        return token;
    }

    public Timestamp getValidUntil() {
        return validUntil;
    }

    public Boolean getInUse() {
        return inUse;
    }

    public void setAsInUse() {
        this.inUse = true;
    }

    public House getHouse() {
        return house;
    }
}

File: gateway_app/db/entities/authentication_token.go
package entities

import (
	"strings"
	"time"

	"github.com/google/uuid"
	"github.com/uptrace/bun"
)

type AuthenticationToken struct {
	bun.BaseModel `bun:"table:gateway.authentication_tokens,alias:at"`

	Id         uuid.UUID `bun:"id,pk,type:uuid,default:uuid_generate_v4()"`
	Code       string    `bun:"token,unique,notnull"`
	ValidFrom  time.Time `bun:"valid_from,notnull"`
	ValidUntil time.Time `bun:"valid_until,notnull"`
	IsValid    bool      `bun:"is_valid,notnull,default:true"`
	InUse      bool      `bun:"in_use,notnull,default:false"`
	HouseId    uuid.UUID `bun:"house_id,type:uuid,notnull"`
	House      *House    `bun:"rel:belongs-to,join:house_id=id"`
	Users      []*User   `bun:"rel:has-many,join:id=id_authentication_token"`
}

func GetNewAuthenticationTokenEntity(houseId uuid.UUID) *AuthenticationToken {
	randomUUID, _ := uuid.NewUUID()
	code := strings.ReplaceAll(randomUUID.String(), "-", "")[0:6]

	authenticationToken := AuthenticationToken{
		Id:         uuid.New(),
		Code:       code,
		ValidFrom:  time.Now(),
		ValidUntil: time.Now().AddDate(1, 0, 0),
		IsValid:    true,
		InUse:      false,
		HouseId:    houseId,
	}
	return &authenticationToken
}

/*

   public AuthenticationToken(String token, House house) {
       var now = Calendar.getInstance();
       var validUntil = Calendar.getInstance();
       validUntil.add(Calendar.YEAR, 1);
       this.id = UUID.randomUUID();
       this.token = token;
       this.validFrom = new Timestamp(now.getTimeInMillis());
       this.validUntil = new Timestamp(validUntil.getTimeInMillis());
       this.isValid = true;
       this.inUse = false;
       this.house = house;
   }
*/
